### PR TITLE
Linting with Ruff

### PR DESCRIPTION
**Pull request by Zeitsperre**
### Pull Request Checklist:
- [ ] This PR addresses an already opened issue (for bug fixes / features)
    - This PR fixes #xyz
- [ ] Tests for the changes have been added (for bug fixes / features)
  - [ ] (If applicable) Documentation has been added / updated (for bug fixes / features)
- [x] CHANGES.rst has been updated (with summary of main changes)
  - [x] Link to issue (:issue:`number`) and pull request (:pull:`number`) has been added

### What kind of change does this PR introduce?

* Removed the configuration blocks in setup.cfg for `pydocstyle` and `pycodestyle`, and removed most of the checking performed by `flake8` (flake8 still needed for `flake8-alphabetize` and  `flake8-rst`).
* Updated `pyproject.toml` with Ruff-based equivalent linting configuration
* Updated `tox.ini` and `Makefile` to use Ruff. Reordered order of linting calls to trigger more typical violations first.
* Updated dependencies.

### Does this PR introduce a breaking change?

Yes. Ruff is now a dev dependency, while `pycodestyle` and `pydocstyle` are no longer required. The checks that were being performed by those libraries are now redundant thanks to the new configuration.

### Other information:

Ruff is a linter for Python (built with Rust) that is blazingly fast and has been in development for around 1.5 years. [Ruff v0.1.0 was released today](https://astral.sh/blog/ruff-v0.1.0).

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -210,3 +210,71 @@ markers = [
   "requires_docs: mark tests that can only be run with documentation present (deselect with '-m \"not requires_docs\"')",
   "requires_internet: mark tests that require internet access (deselect with '-m \"not requires_internet\"')"
 ]
+
+[tool.ruff]
+src = ["xclim"]
+line-length = 150
+target-version = "py38"
```
Comment (coxipi) [inline pyproject.toml:217]: why specify "target-version" here but this is dropped in  "nbqa-black"? Just not sure when this is needed or not

File: setup.cfg [hunk 2]
```diff
@@ -18,40 +18,27 @@ search = __version__ = "{current_version}"
...
 [flake8]
-exclude = 
+exclude =
 	.git,
 	docs,
 	build,
 	.eggs,
 	tests
-max-line-length = 88
-max-complexity = 12
```
Comment (coxipi) [inline setup.cfg:28]: is removing this related to "ruff" or some independent change?

File: tests/test_bootstrapping.py [hunk 3]
```diff
@@ -1,10 +1,7 @@
 from __future__ import annotations
 
-from typing import Callable
-
```
Comment (coxipi) [inline tests/test_bootstrapping.py:4]: This and many changes below, I don't get what this has to do with formatting. Are those unrelated changes you add in the same PR?

File: setup.cfg (hunk 2 above)
Comment (Zeitsperre) [reply setup.cfg:28]: Yeah, this part is weird: so we had a max-line-length set to 88 (which is needed to conform with black) but we regularly broke that rule all over the place. I opted for us to adopt 150 (black formatting will try to aim for 88 all the same), where lines longer than that are considered failures (or they're explicitly ignored, like in the JSON translations)

The Ruff equivalent tool for looking at complexity is called `mccabe`. This configuration was also moved to the `pyproject.toml` file.

Comment (coxipi) [reply setup.cfg:28]: Ah ok, "flake" needs to do less things that are being done by "ruff" now, correct?

File: pyproject.toml (hunk 1 above)
Comment (Zeitsperre) [reply pyproject.toml:217]: I'm not entirely sure what the target-version here impacts, but it should affect any hooks from Ruff that require a minimum supported Python version.

The configuration options for black now live in the `pyproject.toml` (and have been for a few months now). If the CLI flags for black are not set manually, `nbqa-black` will use that configuration by default.

File: tests/test_bootstrapping.py (hunk 3 above)
Comment (Zeitsperre) [reply tests/test_bootstrapping.py:4]: These are a few changes that were raised by Ruff (unused imports, naming, etc.)

File: xclim/analog.py [hunk 4]
```diff
@@ -289,7 +286,7 @@ def zech_aslan(x: np.ndarray, y: np.ndarray, *, dmin: float = 1e-12) -> float:
     :math:`r \to 0`, it is controllable through the `dmin` parameter.
 
     This version corresponds the :math:`D_{ZAE}` test of :cite:t:`grenier_assessment_2013` (eq. 7), which is a version
-    of :math:`\phi_{NM}` from :cite:t:`aslan_new_2003`, modified by using the standardized  euclidean distance, the log
+    of :math:`\phi_{NM}` from :cite:t:`aslan_new_2003`, modified by using the standardized Euclidean distance, the log
     weight function and choosing :math:`d_{min} = 10^{-12}`.
```
Comment (coxipi) [inline xclim/analog.py:290]: Do we want Ruff to manage this kind of stuff?

**Comment by Zeitsperre** [comment]
For additional context, we still need to use `flake8` until either `ruff` develops some new hooks or a better tool comes along that can deal with RestructuredText formatting (comoplicated) and alphabetical listings in `__all__` (less complicated; see: https://github.com/astral-sh/ruff/issues/1198)

File: xclim/analog.py (hunk 4 above)
Comment (Zeitsperre) [reply xclim/analog.py:290]: Oh, that was me seeing a typo. Ruff isn't that advanced! haha

File: xclim/core/formatting.py [hunk 5]
```diff
@@ -247,8 +247,9 @@ def parse_doc(doc: str) -> dict[str, str]:
 
 
 def _parse_parameters(section):
-    """Parse the 'parameters' section of a docstring into a dictionary mapping the parameter name to its description and, potentially, to its set of choices.
+    """Parse the 'parameters' section of a docstring into a dictionary.
 
+    Works by mapping the parameter name to its description and, potentially, to its set of choices.
     The type annotation are not parsed, except for fixed sets of values (listed as "{'a', 'b', 'c'}").
```
Comment (coxipi) [inline xclim/core/formatting.py:253]: here again, it's nice I agree with in this specific case, but seems a bit powerful. It seems that this could lead to confusion. If I have documentation that is not OK with syntax, but understandable, this could change things and make things less understandable?

File: xclim/core/indicator.py [hunk 6]
```diff
@@ -62,7 +65,7 @@
         allowed_periods: [<list>, <of>, <allowed>, <periods>]
 
         # Compute function
-        compute: <function name>  # Referring to a function in the supplied `Indices` module, xclim.indices.generic or xclim.indices
+        compute: <function name>  # Referring to a function in `Indices` module (xclim.indices.generic or xclim.indices)
         input:  # When "compute" is a generic function, this is a mapping from argument name to the expected variable.
```
Comment (coxipi) [inline xclim/core/indicator.py:69]: if Ruff is responsible for this removal, it is being quite opiniated, no?

File: xclim/core/formatting.py (hunk 5 above)
Comment (Zeitsperre) [reply xclim/core/formatting.py:253]: That was a manual change by me to bring the function title down to a reasonable length. If this changes the meaning entirely, I can revert that

File: xclim/core/indicator.py (hunk 6 above)
Comment (coxipi) [reply xclim/core/indicator.py:69]: That is stylistic IMO

Comment (Zeitsperre) [reply xclim/core/indicator.py:69]: Nope, again, this was my doing.

File: xclim/indices/_agro.py [hunk 7]
```diff
@@ -977,12 +982,13 @@ def rain_season(
...
-    on `method_dry_start`). The rain season stops when a dry sequence happens (the dry sequence is defined as in the start sequence,
-    but with parameters `window_dry_end`, `thresh_dry_end` and `method_dry_end`). The dates on which the season can start are constrained
-    by `date_min_start`and `date_max_start` (and similarly for the end of the season).
+    The rain season starts at the end of a period of raining (a total precipitation  of `thresh_wet_start` over
+    `window_wet_start` days). This must be directly followed by a period of `window_not_dry_start` days with no dry
+    sequence. The dry sequence is a period of `window_dry_start` days where precipitations are below `thresh_dry_start`
+    (either the total precipitations over the period, or the daily precipitations, depending on `method_dry_start`).
+    The rain season stops when a dry sequence happens (the dry sequence is defined as in the start sequence, but with
+    parameters `window_dry_end`, `thresh_dry_end` and `method_dry_end`). The dates on which the season can start are
+    constrained by `date_min_start`and `date_max_start` (and similarly for the end of the season).
```
Comment (coxipi) [inline xclim/indices/_agro.py:992]: Ruff can remove the job of making docstring lines not too long? That's very nice.

Comment (Zeitsperre) [reply xclim/indices/_agro.py:992]: Ruff is more suited to letting people know when lines are too long. There are a few types of fixes considered `safe` that it can do though: https://docs.astral.sh/ruff/configuration/#fix-safety

**Comment by coxipi** [comment]
Ok, I may have mixed manual and Ruff changes, you can tell me.  

It doesn't seem to be much different than what we had before. Sometimes more/less white space. It seems good. I'm not sure what could go wrong with PR really, what should I be looking for. It didn't change much of our earlier formatting, so it must be working more or less equivalently.

File: tests/test_bootstrapping.py (hunk 3 above)
Comment (coxipi) [reply tests/test_bootstrapping.py:4]: Ok, not sure about this one. Say you are working on a PR and you comment the only line that depends on an import. Then, later on, if you commit, then you want to uncomment the line, you would need to rewrite the import that was formatted out? Or will this take into consideration commented lines too?

**Comment by coxipi** [comment]
need to remove mentions of pydocstyle in "contributing.rst" I think?

File: tests/test_bootstrapping.py (hunk 3 above)
Comment (coxipi) [reply tests/test_bootstrapping.py:4]: I'd be ready to give my go, just wanted to know what you think of this comment before

**Review by coxipi — APPROVED** [review]
Approved!

File: tests/test_bootstrapping.py (hunk 3 above)
Comment (Zeitsperre) [reply tests/test_bootstrapping.py:4]: We now have much more enforced checks for unused imports. Essentially, if you have an unused import outside a `__init__.py` file, the checker will complain until you either remove it or add a `# noqa: F401` to the import line.
